Limit threads

eviefp opened the pull request:
This is more of a "is this what you had in mind", I'm totally open to feedback.

I added a `--threads N` option for all commands that (eventually) hit a `mapConcurrently` or a `forConcurrently_`. I did it by adding a `Maybe Int` parameter down the call stack (not sure if it's worth it refactoring all of the functions to some `ReaderT env IO`; probably not with this PR but I could take that on if it would get accepted).

A value of `Nothing` means use old behaviour, otherwise use the method suggested by @kRITZCREEK's link in #126.

I would also like to keep the changes to the import because I like them arranged nicely, but I'm totally fine reverting that chunk if you want me to.

This has been tested on Appveyor on our project and fixed our issues. We have only tested with `--threads 1` so far and that works (but is obviously a bit slow).

Commit:
moving

Commit:
test

Commit:
fix?

Commit:
whoops

Commit:
limitThreads all the things

justinwoo commented on app/Main.hs:
```diff
@@ -586,6 +605,10 @@ main = do
 
+        limitThreads = Opts.option Opts.auto $
+             Opts.long "threads"
+          <> Opts.help "Limit the number of threads"
```
Needs better explanation

eviefp replied:
How about renaming it to `jobs` and the text to _Limit the number of jobs that can run concurrently_?

justinwoo commented on app/Main.hs:
```diff
@@ -1,46 +1,52 @@
-import           System.Environment (getArgs)
-import qualified System.IO as IO
-import qualified System.Process as Process
+import           Data.Either.Combinators      (rightToMaybe)
```
I really would rather not have all of these diff changes because it makes maintenance harder

justinwoo commented:
This sounds fine overall? Just I guess we should do a pre-release in any case if we merge this to see if anyone has problems with this.

eviefp commented:
No rush for me, we've included the binary from this branch in our repo until it's up on `chocolatey`.

kritzcreek commented on psc-package.cabal:
```diff
@@ -27,7 +27,8 @@ executable psc-package
                    errors -any,
-                   turtle <1.6
+                   turtle <1.6,
+                   SafeSemaphore -any
```
No need to depend on this package, the bugs in base's semaphore have been fixed.

kritzcreek reviewed (changes requested):
I don't think we should be using the word thread, it's jobs not threads that we are limiting here (so the word thread might mislead someone).

Commit:
addressed reviews, still need to ttest

Commit:
removed two more formatting changes

eviefp commented:
Thanks for the reviews, I think I addressed the issues. Let me know if the current message is okay @justinwoo 

I haven't tested it yet, will do that tomorrow at work on our codebase and will update then.

Commit:
forgot to rename argument to jobs

Commit:
signalQSem

eviefp commented:
Alright, after a bit of confusion I think everything is working correctly. Tested it on appveyor and it works just fine with up to 6 jobs. Didn't encounter any issues with this version.

kritzcreek commented on app/Main.hs:
```diff
@@ -505,7 +516,11 @@ verify arg = do
+        Nothing -> mapConcurrently dirFor dependencies
+        Just max' -> do
+          sem <- newQSem max'
+          mapConcurrently (\x -> waitQSem sem *> dirFor x <* signalQSem sem) dependencies
```
You'll want to use `bracket_ (waitQSem sem) (signalQSem sem) (dirFor x)` to make sure you don't lose a capacity on exception in `dirFor`.

EDIT:

I got that one from http://hackage.haskell.org/package/base-4.11.1.0/docs/Control-Concurrent-QSem.html#t:QSem

eviefp replied:
Good catch, thank you!

Commit:
should use bracket_, removed the styling changes again

Commit:
couple of extra spacing causing noise in the diff

Commit:
and one more

eviefp commented:
Added `bracket_`, made sure there's no more formatting noise and tested it on our appveyor. Seems to work even with jobs = 10, but still fails with unlimited.

If you mind the Makefile I could remove that too.

justinwoo commented:
Thanks! I made a pre-release here: https://github.com/purescript/psc-package/releases/tag/v0.4.1-pre
